<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Manual</title>
  <meta name="description" content="Manual">
  <meta name="generator" content="bookdown 0.6 and GitBook 2.6.7">

  <meta property="og:title" content="Manual" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Manual" />
  
  
  

<meta name="author" content="Sarah Williams">


<meta name="date" content="2018-02-21">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="background.html">
<link rel="next" href="using-the-package.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="background.html"><a href="background.html"><i class="fa fa-check"></i><b>1</b> Background</a></li>
<li class="chapter" data-level="2" data-path="overview.html"><a href="overview.html"><i class="fa fa-check"></i><b>2</b> Overview</a><ul>
<li class="chapter" data-level="2.1" data-path="overview.html"><a href="overview.html#workflow"><i class="fa fa-check"></i><b>2.1</b> Workflow</a><ul>
<li class="chapter" data-level="2.1.1" data-path="overview.html"><a href="overview.html#prepare-dataset"><i class="fa fa-check"></i><b>2.1.1</b> Prepare dataset</a></li>
<li class="chapter" data-level="2.1.2" data-path="overview.html"><a href="overview.html#within-dataset-de-differential-expression"><i class="fa fa-check"></i><b>2.1.2</b> Within dataset DE (Differential Expression)</a></li>
<li class="chapter" data-level="2.1.3" data-path="overview.html"><a href="overview.html#query-ref-comparison"><i class="fa fa-check"></i><b>2.1.3</b> Query-Ref Comparison</a></li>
<li class="chapter" data-level="2.1.4" data-path="overview.html"><a href="overview.html#interpreting-output"><i class="fa fa-check"></i><b>2.1.4</b> Interpreting output</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="overview.html"><a href="overview.html#assigning-labels-to-clusters"><i class="fa fa-check"></i><b>2.2</b> Assigning labels to clusters</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="using-the-package.html"><a href="using-the-package.html"><i class="fa fa-check"></i><b>3</b> Using the package</a><ul>
<li class="chapter" data-level="3.1" data-path="using-the-package.html"><a href="using-the-package.html#quickstart"><i class="fa fa-check"></i><b>3.1</b> Quickstart</a><ul>
<li class="chapter" data-level="3.1.1" data-path="using-the-package.html"><a href="using-the-package.html#example-simulated-data-100-genes"><i class="fa fa-check"></i><b>3.1.1</b> Example: Simulated data, 100 genes</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="using-the-package.html"><a href="using-the-package.html#Input"><i class="fa fa-check"></i><b>3.2</b> Input</a><ul>
<li class="chapter" data-level="3.2.1" data-path="using-the-package.html"><a href="using-the-package.html#loading-data"><i class="fa fa-check"></i><b>3.2.1</b> Loading data</a></li>
<li class="chapter" data-level="3.2.2" data-path="using-the-package.html"><a href="using-the-package.html#extra-data-preparation"><i class="fa fa-check"></i><b>3.2.2</b> Extra data preparation</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="using-the-package.html"><a href="using-the-package.html#Prepare-data-with-within-experiment-differential-expression"><i class="fa fa-check"></i><b>3.3</b> Prepare data with within-experiment differential expression</a></li>
<li class="chapter" data-level="3.4" data-path="using-the-package.html"><a href="using-the-package.html#making-comparisons-to-reference-data"><i class="fa fa-check"></i><b>3.4</b> Making comparisons to reference data</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="example-analyses.html"><a href="example-analyses.html"><i class="fa fa-check"></i><b>4</b> Example Analyses</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Manual</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="overview" class="section level1">
<h1><span class="header-section-number">2</span> Overview</h1>
<div id="workflow" class="section level2">
<h2><span class="header-section-number">2.1</span> Workflow</h2>
<p><img src="images/workflow_diagram.png" alt="Workflow of diagram " /> To characterise a query dataset’s cell clusters on the basis of transcriptomic similarity to a annotated reference dataset, the workflow is described in figure .</p>
<div id="prepare-dataset" class="section level3">
<h3><span class="header-section-number">2.1.1</span> Prepare dataset</h3>
<p>The input required is a counts matrix of number of reads per gene per cell, and the cluster assignment for each cell. Cell clusters might be defined by any cell-clustering technique, such as those implemented in tools such as Seurat[cite], cellRanger (10X genomics), SC3[cite], among many others.</p>
</div>
<div id="within-dataset-de-differential-expression" class="section level3">
<h3><span class="header-section-number">2.1.2</span> Within dataset DE (Differential Expression)</h3>
<p>Every dataset, whether a query or a reference, is prepared the same way. For every cluster, cells within that cluster are compared to the rest of the cells pooled together, calculating differential gene expression using MAST [cite]. Because of the low counts and potential dropout issues in single cell RNAseq data, only genes enriched in each cluster are considered. For every cluster – cells are ranked from most to least enriched according to their lower 95% CI of fold-change[cite]. Each gene is assigned a ‘rescaled rank’ from 0 (most enriched) to 1. This step is the most time consuming, but only needs to be done once per dataset.</p>
</div>
<div id="query-ref-comparison" class="section level3">
<h3><span class="header-section-number">2.1.3</span> Query-Ref Comparison</h3>
<p>A list of ‘Up’ genes are extracted for each query cluster – defined as those that have significantly higher expression in that cluster versus the rest of the sample (p&lt;0.01 after BH multiple hypothesis correction). The ‘Up’ gene list is capped at the top 100 (ranked by lower 95% FC). Then, those genes are looked up in the ranking of genes in each reference cell cluster. The distribution of these ‘up gene’ ranks is plotted to evaluate similarity of the query cell-group to a reference cell-group.</p>
</div>
<div id="interpreting-output" class="section level3">
<h3><span class="header-section-number">2.1.4</span> Interpreting output</h3>
<p>Typically, every cell cluster in the query data (each box) is plotted against everything in the reference data (X-axis). Each of the ‘up’ genes is represented by a tick mark, and a the median of ranks is shown as a thick bar. A bias near the top indicates similarity of the groups – essentially the same genes are representative of the clusters within their respective samples. The dotted line at 0.25 is an arbitrary threshold of similarity (the top quartile of genes). Note that 0.5 indicates a completely random distribution, but lower values are common due to the reciprocal nature of the within-dataset differential expression.</p>
</div>
</div>
<div id="assigning-labels-to-clusters" class="section level2">
<h2><span class="header-section-number">2.2</span> Assigning labels to clusters</h2>
<p><img src="images/clustering_labelling.png" alt="Cluster labelling " /> This package suggests some semi-sensible query cluster labels on the basis of similarity to the annotated reference clusters. A rough indication of similarity is when the median gene rank (thick bar) is less than (plotted above) the arbitrary threshold of similarity at 0.25 (dotted line). Because some cell populations are more distinct then others, so some further logic is applied:</p>
<ol style="list-style-type: upper-alpha">
<li>If only one reference group is above the threshold, use that group name.</li>
<li>If nothing meets the threshold, report a lack of similarity.</li>
<li>If two (or more) reference groups are above the threshold, run a Kolmogorov–Smirnov test on the rescaled rank distributions. If they are significantly (p &lt;= 0.01) different (which might occur for similar but distinct cell types), report only the first.</li>
<li>Or if they are not, report both. If there are more than two groups above the threshold, keep testing the first to the second, second to third e.t.c, until a significant difference is found, or there are no more above-threshold groups, reporting everything seen up to that point.</li>
</ol>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="background.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="using-the-package.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
},
"search": false
});
});
</script>

</body>

</html>
