<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Manual</title>
  <meta name="description" content="Manual">
  <meta name="generator" content="bookdown 0.6 and GitBook 2.6.7">

  <meta property="og:title" content="Manual" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Manual" />
  
  
  

<meta name="author" content="Sarah Williams">


<meta name="date" content="2018-02-21">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="overview.html">
<link rel="next" href="example-analyses.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="background.html"><a href="background.html"><i class="fa fa-check"></i><b>1</b> Background</a></li>
<li class="chapter" data-level="2" data-path="overview.html"><a href="overview.html"><i class="fa fa-check"></i><b>2</b> Overview</a><ul>
<li class="chapter" data-level="2.1" data-path="overview.html"><a href="overview.html#workflow"><i class="fa fa-check"></i><b>2.1</b> Workflow</a><ul>
<li class="chapter" data-level="2.1.1" data-path="overview.html"><a href="overview.html#prepare-dataset"><i class="fa fa-check"></i><b>2.1.1</b> Prepare dataset</a></li>
<li class="chapter" data-level="2.1.2" data-path="overview.html"><a href="overview.html#within-dataset-de-differential-expression"><i class="fa fa-check"></i><b>2.1.2</b> Within dataset DE (Differential Expression)</a></li>
<li class="chapter" data-level="2.1.3" data-path="overview.html"><a href="overview.html#query-ref-comparison"><i class="fa fa-check"></i><b>2.1.3</b> Query-Ref Comparison</a></li>
<li class="chapter" data-level="2.1.4" data-path="overview.html"><a href="overview.html#interpreting-output"><i class="fa fa-check"></i><b>2.1.4</b> Interpreting output</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="overview.html"><a href="overview.html#assigning-labels-to-clusters"><i class="fa fa-check"></i><b>2.2</b> Assigning labels to clusters</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="using-the-package.html"><a href="using-the-package.html"><i class="fa fa-check"></i><b>3</b> Using the package</a><ul>
<li class="chapter" data-level="3.1" data-path="using-the-package.html"><a href="using-the-package.html#quickstart"><i class="fa fa-check"></i><b>3.1</b> Quickstart</a><ul>
<li class="chapter" data-level="3.1.1" data-path="using-the-package.html"><a href="using-the-package.html#example-simulated-data-100-genes"><i class="fa fa-check"></i><b>3.1.1</b> Example: Simulated data, 100 genes</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="using-the-package.html"><a href="using-the-package.html#Input"><i class="fa fa-check"></i><b>3.2</b> Input</a><ul>
<li class="chapter" data-level="3.2.1" data-path="using-the-package.html"><a href="using-the-package.html#loading-data"><i class="fa fa-check"></i><b>3.2.1</b> Loading data</a></li>
<li class="chapter" data-level="3.2.2" data-path="using-the-package.html"><a href="using-the-package.html#extra-data-preparation"><i class="fa fa-check"></i><b>3.2.2</b> Extra data preparation</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="using-the-package.html"><a href="using-the-package.html#Prepare-data-with-within-experiment-differential-expression"><i class="fa fa-check"></i><b>3.3</b> Prepare data with within-experiment differential expression</a></li>
<li class="chapter" data-level="3.4" data-path="using-the-package.html"><a href="using-the-package.html#making-comparisons-to-reference-data"><i class="fa fa-check"></i><b>3.4</b> Making comparisons to reference data</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="example-analyses.html"><a href="example-analyses.html"><i class="fa fa-check"></i><b>4</b> Example Analyses</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Manual</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="using-the-package" class="section level1">
<h1><span class="header-section-number">3</span> Using the package</h1>
<div id="quickstart" class="section level2">
<h2><span class="header-section-number">3.1</span> Quickstart</h2>
<div id="example-simulated-data-100-genes" class="section level3">
<h3><span class="header-section-number">3.1.1</span> Example: Simulated data, 100 genes</h3>
<p>Suppose there’s a new single-cell RNAseq dataset (query), and the cells have been clustered into 4 groups : Groups 1-4. But we don’t know what they are. Luckily, there’s an older dataset (ref) of the same tissue type in which someone has already determined the cell types.</p>
<p>Use the reference dataset to figure out what type of cells are in each cluster of the new experiment</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
<span class="kw">library</span>(celaref)

<span class="co"># Paths to data files.</span>
counts_filepath.query    &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;sim_query_counts.tab&quot;</span>,    <span class="dt">package =</span> <span class="st">&quot;celaref&quot;</span>)
cell_info_filepath.query &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;sim_query_cell_info.tab&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;celaref&quot;</span>)
counts_filepath.ref      &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;sim_ref_counts.tab&quot;</span>,      <span class="dt">package =</span> <span class="st">&quot;celaref&quot;</span>)
cell_info_filepath.ref   &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;sim_ref_cell_info.tab&quot;</span>,   <span class="dt">package =</span> <span class="st">&quot;celaref&quot;</span>)

<span class="co"># Load data</span>
toy_ref_se   &lt;-<span class="st"> </span><span class="kw">load_se_from_files</span>(<span class="dt">counts_file=</span>counts_filepath.ref, <span class="dt">cell_info_file=</span>cell_info_filepath.ref)
toy_query_se &lt;-<span class="st"> </span><span class="kw">load_se_from_files</span>(<span class="dt">counts_file=</span>counts_filepath.query, <span class="dt">cell_info_file=</span>cell_info_filepath.query)

<span class="co"># Filter data</span>
toy_ref_se     &lt;-<span class="st"> </span><span class="kw">trim_small_groups_and_low_expression_genes</span>(toy_ref_se)
toy_query_se   &lt;-<span class="st"> </span><span class="kw">trim_small_groups_and_low_expression_genes</span>(toy_query_se)

<span class="co"># Setup within-experiment differential expression</span>
de_table.toy_ref   &lt;-<span class="st"> </span><span class="kw">contrast_each_group_to_the_rest</span>(toy_ref_se,    <span class="dt">dataset_name=</span><span class="st">&quot;ref&quot;</span>)
de_table.toy_query &lt;-<span class="st"> </span><span class="kw">contrast_each_group_to_the_rest</span>(toy_query_se,  <span class="dt">dataset_name=</span><span class="st">&quot;query&quot;</span>)

<span class="co"># Cross-experimet comparison</span>
de_table.marked.query_vs_ref &lt;-<span class="st"> </span><span class="kw">get_the_up_genes_for_all_possible_groups</span>(
   <span class="dt">de_table.test=</span>de_table.toy_query ,
   <span class="dt">de_table.ref=</span>de_table.toy_ref ,
   <span class="dt">test_dataset_name=</span><span class="st">&#39;query&#39;</span>)

<span class="co"># Plot</span>
<span class="kw">make_ranking_violin_plot</span>(de_table.marked.query_vs_ref)

<span class="co"># And get group labels</span>
<span class="kw">make_ref_similarity_names_for_groups_ks</span>(de_table.marked.query_vs_ref, <span class="dt">the_test_dataset=</span><span class="st">&#39;query&#39;</span>, <span class="dt">the_ref_dataset =</span> <span class="st">&quot;ref&quot;</span>)</code></pre></div>
<p>The above commands use a toy dataset within the package, and can be copy-pasted and run. It is a tiny simulated dataset that will complete very quickly.</p>
</div>
</div>
<div id="Input" class="section level2">
<h2><span class="header-section-number">3.2</span> Input</h2>
<p>The celaref package works with datasets in ‘SummarizedExperiment’ objects. While they can be constructed ‘by hand’ there are several functions (below) to create them in a format with all the required information.</p>
<p>The following peices of information are needed to use a single cell RNAseq dataset with celaref.</p>
<ol style="list-style-type: decimal">
<li><strong>Counts Matrix</strong> Number of gene tags per gene per cell.</li>
<li><strong>Cell information</strong> A sample-sheet table of cell-level information. Only two fields are essential:
<ul>
<li><em>cell_sample</em>: A unique cell identifier</li>
<li><em>group</em>: The cluster/group to which the cell has been assigned.</li>
</ul></li>
<li><strong>Gene information</strong> Optional. A table of extra gene-level information.
<ul>
<li><em>ID</em>: A unique gene identifier</li>
</ul></li>
</ol>
<p>The cell information tables can contain whatever experimentally relevant data is desired, like treatment, batches, individual …</p>
<p>The celaref package doesn’t do any clustering itself - cells should have already been assigned to cluster groups on the basis of transcriptional similarity using one of the many single-cell clustering tools available. Note that any cells not assigned to a group will not be processed.</p>
<p>For a querying dataset clusters will of course have arbitrary names like c1,c2,c3.., but for reference datasets they should be something meaningful (e.g. ‘macrophages’).</p>
<p>Providing gene-level information is entirely optional, because it can be taken from the counts matrix. It is useful for tracking multiple IDs, e.g. [Converting-IDs]</p>
<div id="loading-data" class="section level3">
<h3><span class="header-section-number">3.2.1</span> Loading data</h3>
<div id="from-tables-or-flat-files" class="section level4">
<h4><span class="header-section-number">3.2.1.1</span> From tables or flat files</h4>
<p>The simplest way to load data is with two files.</p>
<p>A tab-delimited counts matrix formatted as follows:</p>
<table>
<thead>
<tr class="header">
<th>gene</th>
<th>Cell1</th>
<th>cell2</th>
<th>cell3</th>
<th>cell4</th>
<th>…</th>
<th>cell954</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>GeneA</td>
<td>0</td>
<td>1</td>
<td>0</td>
<td>1</td>
<td>…</td>
<td>0</td>
</tr>
<tr class="even">
<td>GeneB</td>
<td>0</td>
<td>3</td>
<td>0</td>
<td>2</td>
<td>…</td>
<td>2</td>
</tr>
<tr class="odd">
<td>GeneC</td>
<td>1</td>
<td>40</td>
<td>1</td>
<td>0</td>
<td>…</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>And a tab-delimited cell info / sample-sheet file of cell-level information, including the group assignment foreach cell (‘Cluster’), and any other useful information.</p>
<table>
<thead>
<tr class="header">
<th>CellId</th>
<th>Sample</th>
<th>Cluster</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>cell1</td>
<td>Control</td>
<td>cluster1</td>
</tr>
<tr class="even">
<td>cell2</td>
<td>Control</td>
<td>cluster7</td>
</tr>
<tr class="odd">
<td>…</td>
<td>…</td>
<td>…</td>
</tr>
<tr class="even">
<td>cell954</td>
<td>KO</td>
<td>cluster8</td>
</tr>
</tbody>
</table>
<p>Internally, (and throughout this doco), there are references to the ‘cell_sample’ and ‘group’ columns. They can be named as such in the input cell info table, or just assumed or specified when loaded.</p>
<p>The following commands are equivalent, because the first column of the cell info table is assumed to be the cell identifier if not specified.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
dataset_se &lt;-<span class="st"> </span><span class="kw">load_se_from_files</span>(<span class="dt">counts_matrix   =</span> <span class="st">&quot;counts_matrix_file.tab&quot;</span>, 
                                  <span class="dt">cell_info_file =</span> <span class="st">&quot;cell_info_file.tab&quot;</span>, 
                                  <span class="dt">group_col_name  =</span> <span class="st">&quot;Cluster&quot;</span>)

dataset_se &lt;-<span class="st"> </span><span class="kw">load_se_from_files</span>(<span class="dt">counts_matrix   =</span> <span class="st">&quot;counts_matrix_file.tab&quot;</span>, 
                                  <span class="dt">cell_info_file =</span> <span class="st">&quot;cell_info_file.tab&quot;</span>,
                                  <span class="dt">group_col_name  =</span> <span class="st">&quot;Cluster&quot;</span>,
                                  <span class="dt">cell_col_name   =</span> <span class="st">&quot;CellId&quot;</span> )</code></pre></div>
<p>If there was a table of extra information on the genes:</p>
<table>
<thead>
<tr class="header">
<th>Gene</th>
<th>NiceName</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>GeneA</td>
<td>NiceNameA</td>
</tr>
<tr class="even">
<td>GeneB</td>
<td>NiceNameB</td>
</tr>
<tr class="odd">
<td>GeneC</td>
<td>NiceNameC</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dataset_se &lt;-<span class="st"> </span><span class="kw">load_se_from_files</span>(<span class="dt">counts_matrix   =</span> <span class="st">&quot;counts_matrix_file.tab&quot;</span>, 
                                  <span class="dt">cell_info_file =</span> <span class="st">&quot;cell_info_file.tab&quot;</span>, 
                                  <span class="dt">gene_info_file =</span> <span class="st">&quot;gene_info_file.tab&quot;</span>,
                                  <span class="dt">group_col_name =</span> <span class="st">&quot;Cluster&quot;</span>)</code></pre></div>
<p>If extra gene information is included, the first column (or a column named ID) must be unique. Every gene must match between the counts matrix and gene info.</p>
<p>In contrast, if cell information is missing (from cell info or from the counts), the cell will just be dropped from the analysis. This is useful when excluding cells or subsetting the analysis - it is enough to remove entries from the cell info table. If this happens, a warning message will display the number of cells kept.</p>
<p>If the data is already loaded into R, the <em>load_se_from_tables</em> funtion will accept data frames instead of filenames.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dataset_se &lt;-<span class="st"> </span><span class="kw">load_se_from_tables</span>(<span class="dt">counts_matrix   =</span> <span class="st">&quot;counts_matrix_file.tab&quot;</span>, 
                                  <span class="dt">cell_info_table =</span> <span class="st">&quot;cell_info_file.tab&quot;</span>, 
                                  <span class="dt">group_col_name  =</span> <span class="st">&quot;Cluster&quot;</span>)</code></pre></div>
</div>
<div id="from-10x-pipeline-output" class="section level4">
<h4><span class="header-section-number">3.2.1.2</span> From 10X pipeline output</h4>
<p>The 10X cellRanger pipelines produce a directory of output including the counts matrix files and several different clusters. This kind of output directory will contain sub-directories called ‘analysis’, ‘filtered_gene_bc_matrices’</p>
<p>To read in a human (GRCh38) dataset using the ‘kmeans_7_clusters’ clustering:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dataset_se &lt;-<span class="st"> </span><span class="kw">load_dataset_10Xdata</span>(<span class="st">&#39;~/path/to/data/10X_mydata&#39;</span>, 
                                   <span class="dt">dataset_genome =</span> <span class="st">&quot;GRCh38&quot;</span>, 
                                   <span class="dt">clustering_set =</span> <span class="st">&quot;kmeans_7_clusters&quot;</span>) </code></pre></div>
<p>Note that the cell ranger pipelines seem to produce many different cluster sets, their names should be seen in the cell loupe broswer, or listed in the 10X_mydata/analysis/clustering directory.</p>
<p>NB: This function assumed the file at 10X_mydata/filtered_gene_bc_matrices/GRCh38/genes.csv will have columns <ensemblID><GeneSymbol>. See function doco if not the case.</p>
</div>
<div id="special-case-directly-with-summarizedexperiment-objects" class="section level4">
<h4><span class="header-section-number">3.2.1.3</span> Special case: Directly with SummarizedExperiment objects</h4>
<p>The data loding functions here are just convenient ways of making the SummarizedExperiment objects with the content that celaref functions expect, handling naming and checking uniqueness e.t.c. See <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">SummarizedExperiment doco</a></p>
<p>The minimum mandatory feilds are described in <a href="using-the-package.html#Input">Input</a>, specifically:</p>
<ul>
<li>Within colData(dataset_se): <strong>cell_sample</strong> and <strong>group</strong> columns.</li>
<li>Within rowData(dataset_se): <strong>ID</strong> column.</li>
</ul>
<p>Note that <strong>group</strong> needs to be a factor, but <strong>cell_sample</strong> and <strong>ID</strong> should <em>not</em> be factors.</p>
<p>The colData (cell information) and rowData (gene information) should exactly match the columns and rows of the counts matrix.</p>
<p>The counts matrix should be integer counts, and a matrix (sparse matrix not currently supported).</p>
</div>
<div id="Reading-microarray-data" class="section level4">
<h4><span class="header-section-number">3.2.1.4</span> Reading microarray data</h4>
<p>Microarray datasets of purified celltypes can be used as references too. However, the analysis doesn’t use summarisedExperiment objects the same way, so it does the within-experiment differential expression directly.</p>
<p>Refer to [Prepare-data-with-within-experiment-differential-expression] for details.</p>
<p><a href="http://bioconductor.org/packages/release/bioc/html/limma.html">Limma</a> needs to be installed to use this function. Limma is used to calculate the differential expression on the microarrays, rather than MAST which is used for the single-cell RNAseq data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
de_table.microarray &lt;-<span class="st"> </span><span class="kw">contrast_each_group_to_the_rest_for_norm_ma_with_limma</span>(
    <span class="dt">norm_expression_table=</span>demo_microarray_expr, 
    <span class="dt">sample_sheet_table=</span>demo_microarray_sample_sheet,
    <span class="dt">dataset_name=</span><span class="st">&quot;DemoSimMicroarrayRef&quot;</span>, 
    <span class="dt">sample_name=</span><span class="st">&quot;cell_sample&quot;</span>, <span class="dt">group_name=</span><span class="st">&quot;group&quot;</span>) </code></pre></div>
</div>
</div>
<div id="extra-data-preparation" class="section level3">
<h3><span class="header-section-number">3.2.2</span> Extra data preparation</h3>
<div id="filtering-low-expression-genes-cells" class="section level4">
<h4><span class="header-section-number">3.2.2.1</span> Filtering low-expression genes, cells</h4>
<p>It is standard practice to remove uninformative low-expression genes before calculating differential expression. And in single-cell sequencing, low counts can indicate a problem cell - which can be dropped. Similarly, for the celaref package, very small cell groups will not have the statistical power to detect similarity.</p>
<p>The <em>trim_small_groups_and_low_expression_genes</em> function will remove cells and genes that dont meet such thresholds. Defaults are fairly inclusive, and may require tweaking according to different experiments or technologies.</p>
<p>It can be helpful to check the number of genes and cells surviving <em>trim_small_groups_and_low_expression_genes</em> filtering with <em>dim(dataset_se)</em>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Default filtering</span>
dataset_se &lt;-<span class="st"> </span><span class="kw">trim_small_groups_and_low_expression_genes</span>(dataset_se)

<span class="co"># Also defaults, but specified</span>
dataset_se &lt;-<span class="st"> </span><span class="kw">trim_small_groups_and_low_expression_genes</span>(dataset_se, 
                                    <span class="dt">min_lib_size =</span> <span class="dv">1000</span>, 
                                    <span class="dt">min_group_membership =</span> <span class="dv">5</span>,  
                                    <span class="dt">min_detected_by_min_samples =</span> <span class="dv">5</span>)</code></pre></div>
<p>Refer to function doco for exact meaning of these parameters.</p>
</div>
<div id="Converting-IDs" class="section level4">
<h4><span class="header-section-number">3.2.2.2</span> Converting IDs</h4>
<p>Converting one type of gene identfier to another gene identifier is annoying. Even with major identifers like ensembl IDs (ENSG00000139618) or gene symbols (SYN1) there will be imperfect matching (missing ids, multiple matches).</p>
<p>If multiple gene IDs were provided when creating the summarizedExperiment object, (ie a gene info table/file), a convenience function <em>convert_se_gene_ids</em> will allow a graceful conversion between them.</p>
<p>The function needs a tie-breaker for many-to-one gene relationships - picking the one with higher read counts is a decent choice. Note that if both match, the choice is essentially arbitrary (and not neccesarily consistant).</p>
<p>The following code will convert from the original gene IDs (e.g.if ID is ensemblID), to ‘GeneSymbol’ (which should be a column name in rowData(dataset_se))</p>
<ul>
<li>Remove any genes that have no GeneGymbol associated with the ensembl ID</li>
<li>If the same GeneSymbol is assigned to multiple ensembl IDs - it looks up the <em>eval_col</em> value (here specified as ‘total_count’) and picks the bigger one.</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Count and store total reads/gene.</span>
<span class="kw">rowData</span>(dataset_se)<span class="op">$</span>total_count &lt;-<span class="st"> </span><span class="kw">rowSums</span>(<span class="kw">assay</span>(dataset_se))
<span class="co"># rowData(dataset_se) must already list column &#39;GeneSymbol&#39;</span>
dataset_se &lt;-<span class="st"> </span><span class="kw">convert_se_gene_ids</span>(dataset_se, <span class="dt">new_id=</span><span class="st">&#39;GeneSymbol&#39;</span>, <span class="dt">eval_col =</span> <span class="st">&#39;total_count&#39;</span>)</code></pre></div>
<p>It can be helpful to check the number of genes before and after <em>convert_se_gene_ids</em> with <em>dim(dataset_se)</em>.</p>
</div>
</div>
</div>
<div id="Prepare-data-with-within-experiment-differential-expression" class="section level2">
<h2><span class="header-section-number">3.3</span> Prepare data with within-experiment differential expression</h2>
<p>See [Reading-microarray-data]</p>
</div>
<div id="making-comparisons-to-reference-data" class="section level2">
<h2><span class="header-section-number">3.4</span> Making comparisons to reference data</h2>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="overview.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="example-analyses.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
},
"search": false
});
});
</script>

</body>

</html>
